Correct flaws in arc_summary[23] and their test.

rincebrain opened the pull request:
### Motivation and Context
#12036

### Description
So, it turns out arc_summary[23] still have problems like the
one documented in #6965, and the test case didn't notice due to
a flaw.

Redid fixes for both arc_summary scripts, corrected the test case to
fail correctly.

### How Has This Been Tested?
I ran the modified test case on a system without the changes to arc_summary3 and with arc_summary pointing to arc_summary3 and watched it fail. Then I copied the modified version in and watched it pass.

Also tested the arc_summary3 changed on Python 3.6, 3.7, and 3.9.

### Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Library ABI change (libzfs, libzfs\_core, libnvpair, libuutil and libzfsbootenv)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the OpenZFS [code style requirements](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**contributing** document](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md).
- [x] I have added [tests](https://github.com/openzfs/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/openzfs/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

behlendorf reviewed (approved):
Thanks!

rincebrain commented on cmd/arc_summary/arc_summary3:
```diff
@@ -221,6 +222,25 @@ elif sys.platform.startswith('linux'):
+                               err_msg=None, object=None):
+   handle_Exception(exc_type, object, exc_traceback)
+
+def handle_Exception(ex_cls, ex, tb):
+    if ex is IOError:
+        if ex.errno == errno.EPIPE:
+            sys.exit()
+
+    if ex is KeyboardInterrupt:
+        sys.exit()
```
That's correct - the behavior was the same in arc_summary2 with Python 2.

I've pushed a commit that changes that in both arc_summary[23], but I should remark that there are problems with doing that - for example, if I run this on my old (0.8.6) system, I get a KeyError on a missing l2arc kstat - if I raise ex at the end of that block, it prints the error traceback out twice.

rincebrain replied:
That'll teach me not to test things on Python 3.9 again...time to figure out the right way to catch this, sigh.

rincebrain replied:
Okay, there, this has been tested with Python 3.6, 3.7, and 3.9, and verified that it still raises the aforementioned KeyError (albeit printed twice, I think that's non-negotiable) while passing the other checks.

nabijaczleweli commented on cmd/arc_summary/arc_summary3:
```diff
@@ -221,6 +223,32 @@ elif sys.platform.startswith('linux'):
+      os.dup2(devnull,sys.stderr.fileno())
+      # It turns out that while sys.exit() triggers an exception
+      # not handled message on Python 3.8+, os._exit() does not.
+      os._exit(0)
```
this block is misindented

rincebrain replied:
Corrected, thanks!

nabijaczleweli commented on cmd/arc_summary/arc_summary3:
```diff
@@ -221,6 +223,32 @@ elif sys.platform.startswith('linux'):
+        sys.exit()
+
+    if ex_cls is BrokenPipeError:
+      devnull = os.open(os.devnull, os.O_WRONLY)
+      os.dup2(devnull,sys.stdout.fileno())
+      os.dup2(devnull,sys.stderr.fileno())
+      # It turns out that while sys.exit() triggers an exception
+      # not handled message on Python 3.8+, os._exit() does not.
```
You don't need this block (tested on 3.7.3 and 3.9.2) at all (I was also unable to reproduce the error message with sys.exit(), but who knows); the comment may be better served as "Use os._exit() to prevent flushing buffers, which could reraise this" or something maybe

rincebrain replied:
So I don't - it's based on [the suggested way of muting SIGPIPE](https://docs.python.org/3/library/signal.html#note-on-sigpipe), and did indeed work to mute the printed backtrace, though not the error code, from the error I was seeing on Python 3.9.2.

[Here](https://imgur.com/a/oFrAHXo) is a screenshot of what I see if I replace os._exit(0) with sys.exit(0) after commenting out those 3 lines and piping it into head -n 10.

nabijaczleweli replied:
Hm, reading the error you get I do tend to agree WRT the error causation in your comment, but, oddly, the reason that's noted in the documentation *is* sys.exit -> flush -> reraise. Normal day in Python land, I guess?

rincebrain replied:
Who knows? The documentation doesn't necessarily seem inconsistent, since it does say "If an error occurs in the cleanup after the Python interpreter has caught SystemExit (such as an error flushing buffered data in the standard streams), the exit status is changed to 120.", which is indeed what we're seeing here - it's just not that example case.

nabijaczleweli commented on cmd/arc_summary/arc_summary3:
```diff
@@ -221,6 +223,32 @@ elif sys.platform.startswith('linux'):
+
+# It turns out that for Python 3.8+, they reworked how some systems work,
+# such that just using sys.excepthook doesn't solve the problem any more.
+if hasattr(sys,'unraisablehook'): # Python 3.8+
```
missing space after comma, also the EOL-comment seems a bit superfluous

rincebrain replied:
Thanks!

Initially, I added the comment because it was pretty opaque to me (and several people I consulted) what was going on with the initial behavior of apparently ignoring sys.excepthook for a non-unraisable exception, but on reading it again, I agree that it doesn't add much value, so I'll drop it.

rincebrain commented:
Wow, I really mangled that squash. Let's correct that...

jwk404 commented:
Did the re-raise get lost in the squash? It seems like it's still necessary.

rincebrain commented:
It sure did. I'm still confused how I managed to just...drop almost all the commits, as I definitely didn't tell squash to drop anything, but oh well. Should be back now.

rincebrain commented:
Oh boy, I think that CentOS 7 failure is because it installs arc_summary2 and that has a problem. (Based on the facts that it installs arc_summary2 on my CentOS 7 testbed and I can't reproduce such a problem with arc_summary3 on CentOS 7.)

Time for another fishing expedition...

nabijaczleweli commented on cmd/arc_summary/arc_summary3:
```diff
@@ -221,6 +222,28 @@ elif sys.platform.startswith('linux'):
+            sys.exit()
+
+    if ex is KeyboardInterrupt:
+        sys._exit()
```
just noticed this says `sys._exit()` lol

also, the check should be `if ex_cls is KeyboardInterrupt:` to catch this (apparently, tested on 3.7.3 and 3.9.2)

rincebrain replied:
"oops".

I believe it. I even suspect the IOError case would probably need to be ex_cls as well.

nabijaczleweli commented on cmd/arc_summary/arc_summary3:
```diff
@@ -221,6 +222,28 @@ elif sys.platform.startswith('linux'):
+   handle_Exception(exc_type, object, exc_traceback)
+
+def handle_Exception(ex_cls, ex, tb):
+    if ex is IOError:
```
hm, do you have some way to trigger this? i tried and tried, but py3 never gave me IOError(EPIPE) and only gave me BrokenPipeError (unlike py2)

rincebrain replied:
I absolutely do not.

nabijaczleweli replied:
hm, yeah, okay, so in that case i think we should drop the IOError branch. Consider [Python 3's BrokenPipeError](https://docs.python.org/3/library/exceptions.html#BrokenPipeError):
> A subclass of ConnectionError, raised when trying to write on a pipe while the other end has been closed, or trying to write on a socket which has been shutdown for writing. Corresponds to errno EPIPE and ESHUTDOWN.

And compare it to [Python 2's IOError](https://docs.python.org/2/library/exceptions.html#exceptions.IOError):
> Raised when an I/O operation (such as a print statement, the built-in open() function or a method of a file object) fails for an I/O-related reason, e.g., “file not found” or “disk full”.
>
> This class is derived from EnvironmentError.
> The base class for exceptions that can occur outside the Python system: IOError, OSError. When exceptions of this type are created with a 2-tuple, the first item is available on the instance’s errno attribute (it is assumed to be an error number), and the second item is available on the strerror attribute (it is usually the associated error message). The tuple itself is also available on the args attribute.

And, notably,
> The following exceptions are kept for compatibility with previous versions; starting from Python 3.3, they are aliases of OSError.
> exception EnvironmentError
> exception IOError

So if I'm reading this right, IOError(EPIPE) is a strictly py2 construct, and py3 will only ever raise BrokenPipeError in this case

rincebrain replied:
Sounds logical to me. Done.

Thanks for doing so much digging!

rincebrain commented:
I've spent a little bit fiddling with and looking into this, and my judgment is "this isn't fixable in the CentOS 7 case, and I think it might not be fixable on older Python 3 cases" - I just ran arc_summary2 a great many times on my Debian buster (Python 2.7.16) system, and it sometimes writes that noisy error to stderr, but it never returns nonzero - unlike CentOS 7's Python (2.7.5), which alternates between printing "that noisy error":
```
close failed in file object destructor:
sys.excepthook is missing
lost sys.stderr
```
but returning 0, and printing no error traceback out, but returning 1.

Conversely, I also discovered that on CentOS7's Python3 (3.6.8), even hooking sys.excepthook isn't enough to silence its similar behavior, and it's too old for sys.unraisablehook. I even tried muting SIGPIPE with a signal handler like the documentation warns you not to do (...if you're ever doing socket IO), and it does change the behavior - to dying ~every time with the same error code.

So I've just blacklisted arc_summary_001_pos for Python versions <= the one on CentOS 7 (since I haven't gone spelunking to narrow down the version further, it's possible there's a wider range that it won't work for). I also considered blacklisting Python 3 older than 3.7 or 3.8, but that feels too aggressive - I could be persuaded otherwise, though, probably.

(I could also rework the test to try ignoring certain exit codes and then use some of the imperfect strategies above, but I'm not convinced that's a better option?)

If someone has a magical solution to solve this, I'm all ears.

jwk404 commented:
> If someone has a magical solution to solve this, I'm all ears.

What do you think about abandoning the excepthook completely, and placing `main()` in a try block to catch ^C and broken pipes? This way, you've enhanced the script in the way you initially intended, and the test doesn't need to be added to the "maybe" list.

rincebrain commented:
> > If someone has a magical solution to solve this, I'm all ears.
> 
> What do you think about abandoning the excepthook completely, and placing `main()` in a try block to catch ^C and broken pipes? This way, you've enhanced the script in the way you initially intended, and the test doesn't need to be added to the "maybe" list.

Why do you think that would work any better than the excepthook or unraisablehook?

jwk404 commented:
> > > If someone has a magical solution to solve this, I'm all ears.
> > 
> > 
> > What do you think about abandoning the excepthook completely, and placing `main()` in a try block to catch ^C and broken pipes? This way, you've enhanced the script in the way you initially intended, and the test doesn't need to be added to the "maybe" list.
> 
> Why do you think that would work any better than the excepthook or unraisablehook?

Because I've seen lots of python at a variety of versions that catches a keyboard interrupt, and it's always worked as expected. I've never seen it fail until this mechanism is added to the equation. Granted, I am making an assumption, but it feels like one worth testing at least.

rincebrain commented:
I'll test it again, but the case we're particularly trying to catch here isn't really KeyboardInterrupt, and when I tried a try-except around all of the Python3 main, it definitely didn't catch what we needed.

rincebrain commented:
Unmodified vanilla git:
```
$ ./arc_summary2 | head

------------------------------------------------------------------------
ZFS Subsystem Report				Mon May 17 21:16:35 2021
ARC Summary: (HEALTHY)
	Memory Throttle Count:			0

ARC Misc:
	Deleted:				0
	Mutex Misses:				0
	Eviction Skips:				0
Traceback (most recent call last):
  File "./arc_summary2", line 1173, in <module>
    main()
  File "./arc_summary2", line 1168, in main
    page(Kstat)
  File "./arc_summary2", line 1076, in _tunable_summary
    sys.stdout.write(fmt % (name, tunables[name]))
IOError: [Errno 32] Broken pipe
$ ./arc_summary2 | head

------------------------------------------------------------------------
ZFS Subsystem Report				Mon May 17 21:16:37 2021
ARC Summary: (HEALTHY)
	Memory Throttle Count:			0

ARC Misc:
	Deleted:				0
	Mutex Misses:				0
	Eviction Skips:				0
close failed in file object destructor:
sys.excepthook is missing
lost sys.stderr
[rich@osboxes arc_summary]$ ./arc_summary2 | head

------------------------------------------------------------------------
ZFS Subsystem Report				Mon May 17 21:16:37 2021
ARC Summary: (HEALTHY)
	Memory Throttle Count:			0

ARC Misc:
	Deleted:				0
	Mutex Misses:				0
	Eviction Skips:				0
Traceback (most recent call last):
  File "./arc_summary2", line 1173, in <module>
    main()
  File "./arc_summary2", line 1168, in main
    page(Kstat)
  File "./arc_summary2", line 1076, in _tunable_summary
    sys.stdout.write(fmt % (name, tunables[name]))
IOError: [Errno 32] Broken pipe
```
Modified with try-except around main and the sys.excepthook entirely deleted:
```
$ /mnt/arc_summary2 |head

------------------------------------------------------------------------
ZFS Subsystem Report				Mon May 17 21:18:40 2021
ARC Summary: (HEALTHY)
	Memory Throttle Count:			0

ARC Misc:
	Deleted:				0
	Mutex Misses:				0
	Eviction Skips:				0
close failed in file object destructor:
sys.excepthook is missing
lost sys.stderr
$ grep excepthook /mnt/arc_summary2
$
```

rincebrain commented:
arc_summary3 case, unmodified:
```
$ ./arc_summary3 | head;echo $?

------------------------------------------------------------------------
ZFS Subsystem Report                            Mon May 17 21:24:33 2021
Linux 3.10.0-1160.25.1.el7.x86_64                    2.1.99-161_g96f88fe
Machine: osboxes (x86_64)                            2.1.99-161_g96f88fe

ARC status:                                                      HEALTHY
        Memory throttle count:                                         0

ARC size (current):                                     0.0 %    0 Bytes
Traceback (most recent call last):
  File "./arc_summary3", line 956, in <module>
    main()
  File "./arc_summary3", line 950, in main
    section_calls[section](kstats)
  File "./arc_summary3", line 837, in section_tunables
    print(format_raw_line(key, value))
BrokenPipeError: [Errno 32] Broken pipe
1
$ ./arc_summary3 | head;echo $?

------------------------------------------------------------------------
ZFS Subsystem Report                            Mon May 17 21:24:34 2021
Linux 3.10.0-1160.25.1.el7.x86_64                    2.1.99-161_g96f88fe
Machine: osboxes (x86_64)                            2.1.99-161_g96f88fe

ARC status:                                                      HEALTHY
        Memory throttle count:                                         0

ARC size (current):                                     0.0 %    0 Bytes
Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>
BrokenPipeError: [Errno 32] Broken pipe
120
```
Modified with try-except around the entire main (and nothing else, no hooks):
```
$ /mnt/arc_summary3 |head;echo $?

------------------------------------------------------------------------
ZFS Subsystem Report                            Mon May 17 21:25:26 2021
Linux 3.10.0-1160.25.1.el7.x86_64                    2.1.99-161_g96f88fe
Machine: osboxes (x86_64)                            2.1.99-161_g96f88fe

ARC status:                                                      HEALTHY
        Memory throttle count:                                         0

ARC size (current):                                     0.0 %    0 Bytes
0
$ /mnt/arc_summary3 |head;echo $?

------------------------------------------------------------------------
ZFS Subsystem Report                            Mon May 17 21:25:27 2021
Linux 3.10.0-1160.25.1.el7.x86_64                    2.1.99-161_g96f88fe
Machine: osboxes (x86_64)                            2.1.99-161_g96f88fe

ARC status:                                                      HEALTHY
        Memory throttle count:                                         0

ARC size (current):                                     0.0 %    0 Bytes
Exception ignored in: <_io.TextIOWrapper name='<stdout>' mode='w' encoding='UTF-8'>
BrokenPipeError: [Errno 32] Broken pipe
120
```

So unfortunately, it does not solve this problem.

jwk404 commented:
In local testing of arc_summary3 in a loop, I was able to suppress the ignored exception error message by closing stderr in a finally block.

rincebrain commented:
Does that change the return code, though? Because that's the real problem.

jwk404 commented:
If I call arc_summary3 with a script that does `set -o pipefail` first, then the return code is whatever I passed to `sys.exit()`.

rincebrain commented:
I tried that, and it usually worked, but it occasionally spits out a return code 120 instead.

jwk404 commented:
Thanks for catching that; I didn't let it loop long enough.

There was one other thing I was able to do on 2.7.16, 3.6.9 and 3.9.5 which returned the desired value and didn't print unwanted errors through many iterations. I simply caught IOError, calling sys.exit(0) if errno was 32. The other required part was to change the interpreter line to use `python -u`. 

Unfortunately, I wasn't able to find a way to force the use of unbuffered stderr.

- /usr/bin/env doesn't allow you to pass flags to the interpreter in Linux
- Reopening stderr and assigning it to sys.stderr caused the program to produce no output on stdout
- Replacing sys.stderr with a wrapper class that implements flushing only worked on 3.9
- I couldn't see a way to use PYTHONUNBUFFERED

Know any other tricks to emulate `python -u`?

rincebrain commented:
Could do what some of the scripts/ directory does and rewrite the interpreter line in the "make" step.

nabijaczleweli commented:
I mean, nuclear option, but if you stick it at the top before any other imports then it's still not horribly slow:
```python3
import os
import sys
import io
if isinstance(sys.__stderr__.buffer, io.BufferedWriter):
    os.execv(sys.executable, [sys.executable, "-u"] + sys.argv)
```

rincebrain commented:
It looks like `env` supports -S to split the string following it into independent arguments, which might let us get away with passing -u to the interpreter. Gonna go experiment with that now.

nabijaczleweli commented:
`env -S` isn't portable beyond FreeBSD (oddly) and GNU coreutils – a busybox userland, for example, will not have it.

rincebrain commented:
> `env -S` isn't portable beyond FreeBSD (oddly) and GNU coreutils, unfortunately, and a busybox userland, for example, will not have it.

Fine, I'll test and push the exec hack, and if that breaks sometimes I'll wrap it with a shell script, and if that breaks so help me I'll dig out the newest copy of the _old_ `arc_summary.pl`, modernize it, and open a PR for _that_.

rincebrain commented:
Okay, so:
arc_summary2:
* Ubuntu 21.04 x64 (Python 2.7.18) occasionally spits out sys.excepthook messages but doesn't ever return nonzero in hundreds of runs
* Debian 8 x64 (Python 2.7.9) likewise.
* CentOS 7 x64 (Python 2.7.5) likewise.
arc_summary3:
* Debian 11 x64 (Python 3.9.2) doesn't break in hundreds of runs with the exec hack
* Debian 8 x64 (Python 3.4.2) likewise.
* Ubuntu 21.04 x64 (Python 3.9.4, 3.10.0a7) likewise.
* CentOS 7 x64 (Python 3.6.8) likewise
* FreeBSD 13 x64 (Python 3.7.10) likewise

So I've pushed. PTAL, anyone who wants to.

nabijaczleweli commented:
Hm, was worried that execing after all imports instead of just io+os+sys would be catastrophically slow, but it only loses on the order of 20ms (from 0.098s to get to the `ARGS =` line to 0.079ish), so whatever.

The "# We can't use env -S portably, and we need python3 -u, so..." line could use a short "-u for proper signal/pipe/whatever handling, so" explainer, but I don't see any obvious problems with my (untrained) eye.

rincebrain commented:
Better?

nabijaczleweli commented:
Reads well to me!

jwk404 commented:
Looks good. Thanks for the revisions.
